src/timepicker/TimePickerState.ts

```diff
@@ -1,68 +1,106 @@
-/* eslint-disable @typescript-eslint/no-non-null-assertion */
-import React from "react";
-import { useCompositeState, useDisclosureState } from "reakit";
+import * as React from "react";
+import { useControllableState } from "@chakra-ui/hooks";
+import { PopoverInitialState, usePopoverState } from "reakit";
 import { useTimePickerColumnState } from "./TimePickerColumnState";
+import { useSegmentState } from "../segment-spinner/SegmentState";
 
-export interface TimePickerStateProps {
-  onChange?: (v: Date | null) => void;
-  visible?: boolean;
+export interface TimePickerStateProps extends PopoverInitialState {
   value?: string;
+  defaultValue?: string;
+  onChange?: (v: string) => void;
+  formatOptions?: Intl.DateTimeFormatOptions;
+  placeholderDate?: Date;
 }
 
-export const useTimePickerState = ({
-  onChange,
-  visible = false,
-  value = "0:00",
-}: TimePickerStateProps = {}) => {
-  const disclosure = useDisclosureState({ visible });
+export const useTimePickerState = (props: TimePickerStateProps = {}) => {
+  const {
+    value: initialValue,
+    defaultValue: defaultValueProp,
+    onChange,
+    formatOptions = { timeStyle: "short" },
+    placeholderDate,
+  } = props;
 
-  const [time, setTime] = React.useState<Date | null>(parseTime(value));
+  const [timeProp, setTimeProp] = useControllableState({
+    value: initialValue,
+    defaultValue: defaultValueProp,
+    onChange,
+  });
+  console.log("%c timeProp", "color: #99adcc", timeProp);
+
+  const defaultValue =
+    defaultValueProp && isValidTime(defaultValueProp)
+      ? parseTime(defaultValueProp)
+      : parseTime(`${new Date().getHours()}:${new Date().getMinutes()}`);
+  const time = timeProp == null ? defaultValue : parseTime(timeProp);
 
-  React.useEffect(() => {
-    onChange?.(time);
-  }, [onChange, time]);
+  const setTime = (hour: number, minute: number, meridian: string) => {
+    if (meridian === "PM") {
+      hour = +hour + 12;
+    } else {
+      hour = hour % 12;
+    }
+
+    setTimeProp(`${pad(hour)}:${pad(minute)}`);
+  };
+
+  const [hour, setHourProp] = React.useState(time?.getHours() % 12);
+  const [minute, setMinuteProp] = React.useState(time?.getMinutes());
+  const [meridian, setMeridianProp] = React.useState(
+    time?.getHours() >= 12 ? "PM" : "AM",
+  );
+
+  const setHour = (hour: number) => {
+    setHourProp(hour);
+    setTime(hour, minute, meridian);
+  };
+
+  const setMinute = (minute: number) => {
+    setMinuteProp(minute);
+    setTime(hour, minute, meridian);
+  };
+
+  const setMeridian = (meridianValue: string) => {
+    setMeridianProp(meridianValue);
+    setTime(hour, minute, meridianValue);
+  };
 
   const hourState = useTimePickerColumnState({
-    value: time!.getHours() % 12,
-    onChange: v => {
-      const d = new Date(time?.toISOString() as string);
-      d.setHours(v as number);
-      setTime(d);
-    },
-    onSelection: () => disclosure.hide(),
+    value: hour,
+    onChange: setHour,
```

Comment:
TS Error here

![image](https://user-images.githubusercontent.com/35374649/95426670-e7c82d00-0963-11eb-8685-f390ac173f8d.png)


Because the function signature is different.

Reply:
I am aware. Did that at the last moment yesterday.
I will fix it.